File: product/src/core/planning/engine.ts
import { randomUUID } from "node:crypto";

export const nodeStatuses = ["pending", "running", "success", "failed", "pivoted", "blocked", "waiting_approval"] as const;
export const nodeTypes = ["goal", "strategy", "tactic", "action", "observation", "hypothesis", "pivot"] as const;

export type NodeStatus = (typeof nodeStatuses)[number];
export type NodeType = (typeof nodeTypes)[number];

export type PlanNode = {
  id: string;
  type: NodeType;
  status: NodeStatus;
  description: string;
  agent?: string;
  tool?: string;
  tool_args: Record<string, unknown>;
  parent_id?: string;
  children: string[];
  evidence_ids: string[];
  pivot_reason?: string;
  requires_approval: boolean;
  created_at: string;
  completed_at?: string;
};

export type StrategyInput = {
  type?: NodeType;
  description?: string;
  agent?: string;
  tool?: string;
  tool_args?: Record<string, unknown>;
  requires_approval?: boolean;
  reason?: string;
};

export class AttackTree {
  readonly id = randomUUID();
  readonly nodes = new Map<string, PlanNode>();
  readonly created_at = new Date().toISOString();
  readonly goal: string;
  readonly mode: string;
  readonly target?: string;
  root_id?: string;

  constructor(
    goal: string,
    mode = "ctf_challenge",
    target?: string
  ) {
    this.goal = goal;
    this.mode = mode;
    this.target = target;
  }

  addNode(node: PlanNode) {
    this.nodes.set(node.id, node);
    if (node.parent_id && this.nodes.has(node.parent_id)) {
      this.nodes.get(node.parent_id)?.children.push(node.id);
    }
    this.root_id ??= node.id;
    return node;
  }

  getNextActions() {
    return [...this.nodes.values()].filter((node) => {
      if (node.status !== "pending") {
        return false;
      }
      if (!node.parent_id) {
        return true;
      }
      return this.nodes.get(node.parent_id)?.status === "success";
    });
  }

  getFailedPaths() {
    return [...this.nodes.values()].filter((node) => node.status === "failed");
  }

  markComplete(nodeId: string, success: boolean, pivotReason = "") {
    const node = this.nodes.get(nodeId);
    if (!node) {
      return;
    }
    node.status = success ? "success" : "failed";
    node.completed_at = new Date().toISOString();
    if (!success && pivotReason) {
      node.pivot_reason = pivotReason;
    }
  }

  toJSON() {
    return {
      id: this.id,
      goal: this.goal,
      mode: this.mode,
      target: this.target,
      root_id: this.root_id,
      nodes: Object.fromEntries(this.nodes.entries()),
      created_at: this.created_at
    };
  }
}

export class PlanningEngine {
  createPlan(goal: string, mode: string, target?: string) {
    const tree = new AttackTree(goal, mode, target);
    tree.addNode(createPlanNode({
      type: "goal",
      description: goal,
      agent: "master"
    }));
    return tree;
  }

  expandNode(tree: AttackTree, nodeId: string, strategies: StrategyInput[]) {
    return strategies.map((strategy) => tree.addNode(createPlanNode({
      type: strategy.type ?? "action",
      description: strategy.description ?? "",
      agent: strategy.agent,
      tool: strategy.tool,
      tool_args: strategy.tool_args ?? {},
      parent_id: nodeId,
      requires_approval: strategy.requires_approval ?? false
    })));
  }

  pivot(tree: AttackTree, failedNodeId: string, newStrategy: StrategyInput) {
    const failed = tree.nodes.get(failedNodeId);
    if (!failed) {
      return undefined;
    }
    return tree.addNode(createPlanNode({
      type: "pivot",
      description: `Pivot from: ${failed.description}`,
      agent: newStrategy.agent,
      tool: newStrategy.tool,
      tool_args: newStrategy.tool_args ?? {},
      parent_id: failed.parent_id,
      pivot_reason: newStrategy.reason ?? "previous approach failed"
    }));
  }
}

export function createPlanNode(input: Partial<PlanNode> & Pick<PlanNode, "type" | "description">): PlanNode {
  return {
    id: input.id ?? randomUUID(),
    type: input.type,
    status: input.status ?? "pending",
    description: input.description,
    agent: input.agent,
    tool: input.tool,
    tool_args: input.tool_args ?? {},
    parent_id: input.parent_id,
    children: input.children ?? [],
    evidence_ids: input.evidence_ids ?? [],
    pivot_reason: input.pivot_reason,
    requires_approval: input.requires_approval ?? false,
    created_at: input.created_at ?? new Date().toISOString(),
    completed_at: input.completed_at
  };
}

File: product/core/planning/engine.py
"""
z3gh0ne Planning Engine

The core differentiator: security-domain planning with attack trees,
adaptive strategy, and evidence-driven pivots.

Unlike Claude Code/Codex which do linear step-by-step execution,
this engine builds a tree of possible approaches, executes them
adaptively, and pivots when one path fails.
"""
from __future__ import annotations
from dataclasses import dataclass, field
from enum import Enum
from typing import Optional
from uuid import uuid4
from datetime import datetime, timezone


class NodeStatus(str, Enum):
    PENDING = "pending"
    RUNNING = "running"
    SUCCESS = "success"
    FAILED = "failed"
    PIVOTED = "pivoted"
    BLOCKED = "blocked"
    WAITING_APPROVAL = "waiting_approval"


class NodeType(str, Enum):
    GOAL = "goal"
    STRATEGY = "strategy"
    TACTIC = "tactic"
    ACTION = "action"
    OBSERVATION = "observation"
    HYPOTHESIS = "hypothesis"
    PIVOT = "pivot"


@dataclass
class PlanNode:
    id: str = field(default_factory=lambda: str(uuid4()))
    type: NodeType = NodeType.ACTION
    status: NodeStatus = NodeStatus.PENDING
    description: str = ""
    agent: Optional[str] = None
    tool: Optional[str] = None
    tool_args: dict = field(default_factory=dict)
    parent_id: Optional[str] = None
    children: list[str] = field(default_factory=list)
    evidence_ids: list[str] = field(default_factory=list)
    pivot_reason: Optional[str] = None
    requires_approval: bool = False
    created_at: str = field(default_factory=lambda: datetime.now(timezone.utc).isoformat())
    completed_at: Optional[str] = None

    def to_dict(self) -> dict:
        return {k: v.value if isinstance(v, Enum) else v for k, v in self.__dict__.items()}


@dataclass
class AttackTree:
    """
    An attack tree represents the full space of approaches to a goal.
    Unlike a flat task list, it allows:
    - Multiple strategies for the same goal
    - Fallback paths when one approach fails
    - Evidence-driven pruning and expansion
    - Approval gates at high-risk nodes
    """
    id: str = field(default_factory=lambda: str(uuid4()))
    goal: str = ""
    mode: str = "ctf_challenge"
    target: Optional[str] = None
    nodes: dict[str, PlanNode] = field(default_factory=dict)
    root_id: Optional[str] = None
    created_at: str = field(default_factory=lambda: datetime.now(timezone.utc).isoformat())

    def add_node(self, node: PlanNode) -> PlanNode:
        self.nodes[node.id] = node
        if node.parent_id and node.parent_id in self.nodes:
            self.nodes[node.parent_id].children.append(node.id)
        if not self.root_id:
            self.root_id = node.id
        return node

    def get_next_actions(self) -> list[PlanNode]:
        """Get all actionable nodes whose parents are complete."""
        ready = []
        for node in self.nodes.values():
            if node.status != NodeStatus.PENDING:
                continue
            if node.parent_id is None:
                ready.append(node)
            elif self.nodes.get(node.parent_id, PlanNode()).status == NodeStatus.SUCCESS:
                ready.append(node)
        return ready

    def get_failed_paths(self) -> list[PlanNode]:
        """Find failed nodes that might benefit from a pivot."""
        return [n for n in self.nodes.values() if n.status == NodeStatus.FAILED]

    def mark_complete(self, node_id: str, success: bool, pivot_reason: str = ""):
        if node_id not in self.nodes:
            return
        node = self.nodes[node_id]
        node.status = NodeStatus.SUCCESS if success else NodeStatus.FAILED
        node.completed_at = datetime.now(timezone.utc).isoformat()
        if not success and pivot_reason:
            node.pivot_reason = pivot_reason

    def to_dict(self) -> dict:
        return {
            "id": self.id,
            "goal": self.goal,
            "mode": self.mode,
            "target": self.target,
            "root_id": self.root_id,
            "nodes": {k: v.to_dict() for k, v in self.nodes.items()},
            "created_at": self.created_at,
        }


class PlanningEngine:
    """
    Generates and manages attack trees for security tasks.
    
    Key behaviors that beat Claude Code/Codex:
    1. Generates MULTIPLE strategies, not just one linear plan
    2. Executes in parallel where safe
    3. Pivots on failure instead of giving up
    4. Tracks evidence for every decision
    5. Triggers approval for high-risk actions
    """

    def create_plan(self, goal: str, mode: str, target: str = None, context: dict = None) -> AttackTree:
        tree = AttackTree(goal=goal, mode=mode, target=target)
        root = PlanNode(
            type=NodeType.GOAL,
            description=goal,
            agent="master",
        )
        tree.add_node(root)
        return tree

    def expand_node(self, tree: AttackTree, node_id: str, strategies: list[dict]) -> list[PlanNode]:
        """Expand a node with multiple child strategies/actions."""
        new_nodes = []
        for s in strategies:
            child = PlanNode(
                type=NodeType(s.get("type", "action")),
                description=s.get("description", ""),
                agent=s.get("agent"),
                tool=s.get("tool"),
                tool_args=s.get("tool_args", {}),
                parent_id=node_id,
                requires_approval=s.get("requires_approval", False),
            )
            tree.add_node(child)
            new_nodes.append(child)
        return new_nodes

    def pivot(self, tree: AttackTree, failed_node_id: str, new_strategy: dict) -> PlanNode:
        """When a path fails, create a pivot node with alternative approach."""
        failed = tree.nodes.get(failed_node_id)
        if not failed:
            return None
        pivot_node = PlanNode(
            type=NodeType.PIVOT,
            description=f"Pivot from: {failed.description}",
            agent=new_strategy.get("agent"),
            tool=new_strategy.get("tool"),
            tool_args=new_strategy.get("tool_args", {}),
            parent_id=failed.parent_id,
            pivot_reason=new_strategy.get("reason", "previous approach failed"),
        )
        tree.add_node(pivot_node)
        return pivot_node
